### ISSUE TITLE
PEM certificate extracted from PFX is in the wrong order

### ISSUE THREAD

**Issue by pwiens** [Azure/secrets-store-csi-driver-provider-azure#156]
**What steps did you take and what happened:**
Added a PFX certificate which contains a certificate chain with server, intermediate, and root certificate. The PEM file had the structure of:

PRIVATE KEY
ROOT
INTERMEDIATE
SERVER

**What did you expect to happen:**

According to https://tools.ietf.org/html/rfc4346#section-7.4.2 the certificate order should be

SERVER
INTERMEDIATE
ROOT
KEY

If you run openssl x509 -in downloaded.pem -noout -text on the downloaded certificate you'll see the root cert.

**Anything else you would like to add:**
We should also have the option to download the PFX certificate as is.

**Which access mode did you use to access the Azure Key Vault instance:**
Pod Identity

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Azure Key Vault provider version: 0.0.7:
- Kubernetes version: 1.16.10
- Cluster type: AKS

**Comment by aramase** [comment]
Had a discussion with @pwiens offline. The Azure Key Vault Provider returns the certs in the order in which they were uploaded. Unfortunately, in their scenario, they don't have control over the order of the certificates before upload. 

@pwiens is going to take a stab at a PR for this.

**Comment by aramase** [comment]
@pwiens Did you get a chance to look at this?

**Comment by pwiens** [comment]
I did, I think I have a solution. I've been distracted. I have a prototype
that's almost done. I'm going to use the x509 chain functionality in go to
construct the chain and write it out to a pem file. The current solution is
very wrong. It just grabs the data from the PFX and writes if out in the
order it sees it. We shouldn't trust the order. Also, we should probably
write out the cert and key as separate files.

**Comment by drzdbd1** [comment]
@aramase @pwiens , any updates on the said PR? we're also blocked by this gap in adopting CSI driver.

**Comment by ritazh** [comment]
cc @jlichwa

**Comment by jlichwa** [comment]
Rita,

We have to come up with routing issues to us.
CSI Driver should have “no logic” and reflect Key Vault functionality, if there is an issue in Vault it should be routed to us.
Let’s bring it up on meeting tomorrow.

Jack Lichwa

**Comment by drzdbd1** [comment]
> Rita, We have to come up with routing issues to us. CSI Driver should have “no logic” and reflect Key Vault functionality, if there is an issue in Vault it should be routed to us. Let’s bring it up on meeting tomorrow. Jack Lichwa From: Rita Zhang <notifications@github.com> Sent: Monday, August 24, 2020 6:29 PM To: Azure/secrets-store-csi-driver-provider-azure <secrets-store-csi-driver-provider-azure@noreply.github.com> Cc: Jack Lichwa <Jack.Lichwa@microsoft.com>; Mention <mention@noreply.github.com> Subject: Re: [Azure/secrets-store-csi-driver-provider-azure] PEM certificate extracted from PFX is in the wrong order (#156) cc @jlichwa<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjlichwa&data=02%7C01%7CJack.Lichwa%40microsoft.com%7C6ecf05eedae1400d488208d848964588%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637339157532745605&sdata=piC4FMsU8%2Fez%2FkJfMvaZe9ftyuhrKwV%2FiXOXMLsdkkE%3D&reserved=0> — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fsecrets-store-csi-driver-provider-azure%2Fissues%2F156%23issuecomment-679449222&data=02%7C01%7CJack.Lichwa%40microsoft.com%7C6ecf05eedae1400d488208d848964588%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637339157532755597&sdata=oDFYmUsZp4Xn9wCGLIMK2DLcMgAFD%2FDw1a3DJlhfMX0%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAMQULS3QGM7ZJDPH5GT7JJTSCMHWPANCNFSM4OZYVRIQ&data=02%7C01%7CJack.Lichwa%40microsoft.com%7C6ecf05eedae1400d488208d848964588%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637339157532755597&sdata=MeT0HbsMtP%2BMUFfX0H7pv7k2OqSxHjxCzAg3Fk6k%2BtA%3D&reserved=0>.

Added a bit more context here.

We deploy CSI driver for our customers to sync certificates from key vault. The format expected in the environment is PEM and a lot of customers have PFX certificates stored in their key vault. There is no way we could ask them to recreate certificates in PEM. With key vault api, we could easily manipulate the content to generate PEM from PFX but this is not possible today with CSI driver.

**Comment by jlichwa** [comment]
+@sebansal
@drzdbd1  can you provide more information on how you can manipulate using key vault api and can't do it on AKS side? 
If it is something on client libraries which is possible in akv, we can then get that logic from key vault SDK libraries and assist AKS team.

**Comment by drzdbd1** [comment]
It is in key vault client libraries.

With key vault api, we get the SecretBundle, where .Value is the raw bytes from PFX, then construct the X509Certificate2 object so we would be able to enumerate the cert chains to create the desired PEM format. 

CSI driver would always honor the original format so there's no way to convert PFX to PEM in this case.

**Comment by aramase** [comment]
> CSI driver would always honor the original format so there's no way to convert PFX to PEM in this case.

@drzdbd1 We do provide the option to convert PFX to PEM. In fact, that's the current default behavior. If the `objectFormat: pfx` is set then we return the base64 encoded PFX data as is. The order of the certificates returned is the same order in which the cert chain is uploaded to Key Vault.

**Comment by drzdbd1** [comment]
@aramase Just to clarify in case i was confused. If my cert is stored as PFX in key vault and I set `objectFormat: pem`, can I expect PEM as output with **the correct cert chain order**? Because you also mentioned the order would still be the same in which the cert is uploaded but a PEM format with a PFX order would be invalid.

**Comment by aramase** [comment]
> expect PEM as output with the correct cert chain order

@drzdbd1 You're right! The output is PEM and *order is the same in which the cert is uploaded*. This issue is open so we can discuss the changes for the CSI driver to construct the chain with the correct order irrespective of the uploaded content order.

**Comment by drzdbd1** [comment]
I don't see a point that exporting a cert with invalid PEM format would be of any use. To me this is a bug that needs to be fixed.

I wonder `objectFormat` is needed or not as the format could be a metadata retrieved from key vault, I'd like to have a different config like `outputObjectFormat` to set the output format irrespective of the original format in key vault.

**Comment by ritazh** [comment]
This request has been opened with the Key Vault team: Azure/azure-rest-api-specs#10637

**Comment by kenans** [comment]
@aramase Hi, we also ran into the issue recently. Some of the cert chain order is wrong (wrong means the client cert is not the first one and some softwares assume that). I also understand it's because of the safebag order in the PKCS12 file and the CSI KV driver just honors it by default.

Checking some similar issues under the repo, I found the solution is to propose to the KV team to ensure the order on the download secret API. However, I still feel the scope is under the CSI KV driver rather than the KV team and below are my reasons,

1. In some cases we don't have the control over the original PFX certs in KV. Either they are signed by the integrated CA or they are provided by the customer. Ensuring the safe bag order in the original PKCS12 certs are not always possible. 
2. The problem happens when we load PFX from KV but ask the CSI driver to convert it to PEM. Other libs/tools honor the PFX order by default but they give work arounds so the users can re-order their certs if they want. I quote what `openssl pkcs12` command (converting from PFX to PEM) says about it [in the doc](https://www.openssl.org/docs/man1.1.1/man1/pkcs12.html),

    > If none of the -clcerts, -cacerts or -nocerts options are present then all certificates will be output in the order they appear in the input PKCS#12 files. There is no guarantee that the first certificate present is the one corresponding to the private key. Certain software which requires a private key and certificate and assumes the first certificate in the file is the one corresponding to the private key: this may not always be the case. Using the -clcerts option will solve this problem by only outputting the certificate corresponding to the private key. If the CA certificates are required then they can be output to a separate file using the -nokeys -cacerts options to just output CA certificates.

With that being said, I wonder if the CSI KV driver could at least provide an option to re-order the cert? Currently we just cannot use the PFX cert in many cases, such as the ingresss controller, which reports the MD5 of the public key and private key don't match. 

I look forward to your response.
